util/llpcUtil.h

```diff
@@ -158,42 +163,125 @@ inline uint32_t Log2(
+                        res);
+}
+
+template<typename T>
+uint8_t* FromRawBytes(
+    uint8_t* pRawBytesBegin,
+    uint8_t* pRawBytesEnd,
+    T &res)
+{
+  static_assert(sizeof(uint8_t) == 1, "");
+  return reinterpret_cast<uint8_t*>(
+           FromRawBytes(reinterpret_cast<char*>(pRawBytesBegin),
+                        reinterpret_cast<char*>(pRawBytesEnd), res));
+}
+
+template<typename T>
+const uint8_t*
+FromRawBytes(
+    const uint8_t* pRawBytesBegin,
+    const uint8_t* pRawBytesEnd,
+    T &res)
+{
+  return FromRawBytes(const_cast<uint8_t*>(pRawBytesBegin),
+                      const_cast<uint8_t*>(pRawBytesEnd), res);
+}
+
+// ===================================================================================
+// Converts a FromT value to ToT value performing type punning.
+template<typename ToT, typename FromT>
+ToT TypePunTo(
+    const FromT& from) // [in]
+{
+    static_assert(std::is_trivially_copyable<FromT>::value == true,
+                  "Unsupported type");
+    static_assert(std::is_trivial<ToT>::value == true,
+                  "Unsupported type");
+    static_assert(sizeof(FromT) == sizeof(ToT), "Size mismatch!");
+    ToT result;
+    std::memcpy(&result, &from, sizeof(FromT));
+    return result;
+}
+
 // ===================================================================================
 // Returns the bits of a floating point value as an unsigned integer.
 inline uint32_t FloatToBits(
     float f)        // Float to be converted to bits
 {
-    return (*(reinterpret_cast<uint32_t*>(&f)));
+    return TypePunTo<uint32_t>(f);
 }
 
 // ===================================================================================
 // Finds the expected structure in Vulkan structure chain with the specified info.
 template<class T>
-inline const T* FindVkStructInChain(
-    VkStructureType type,    // Vulkan structure type
-    const void*     pNext)   // Base pointer of Vulkan structure
+const T* FindVkStructInChain(
+    VkStructureType type,     // Vulkan structure type
+    const void*     pStart)   // Base pointer of Vulkan structure
 {
-    struct VkStructHeader
+    if (pStart == nullptr)
     {
-        VkStructureType type;
-        VkStructHeader* pNext;
-    };
+        return nullptr;
+    }
+
+    // Follow the linked-list that looks like the below:
+    //   struct VkStructHeader
+    //   {
+    //      VkStructureType type;
+    //      VkStructHeader* pNext;
+    //   };
 
-    const VkStructHeader* pStructHeader = reinterpret_cast<const VkStructHeader*>(pNext);
-    while(pStructHeader != nullptr)
+    const char* pCurrent = static_cast<const char*>(pStart);
+    while (pCurrent != nullptr)
     {
-        if (pStructHeader->type == type)
-        {
-            break;
-        }
-        else
-        {
-            pStructHeader = pStructHeader->pNext;
-        }
+      VkStructureType foundType;
+      pCurrent = FromRawBytes(pCurrent, pCurrent + sizeof(VkStructureType), foundType);
```

Comment:
VkStructHeader isn't defined by LLPC, we should not assume the layout in it. i.e. please use original code.